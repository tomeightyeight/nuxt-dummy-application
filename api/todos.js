'use strict'

import axios from 'axios'

export default {
  async fetchAll () {
    try {
      return await axios({
        method: 'GET',
        url: 'https://jsonplaceholder.typicode.com/todos?userId=1'
      })
    } catch (e) {
      console.warn(e)
    }
  },

  async fetch (id) {
    try {
      return await axios({
        method: 'GET',
        url: `https://jsonplaceholder.typicode.com/todos/${id}`
      })
    } catch (e) {
      console.warn(e)
    }
  }
}
