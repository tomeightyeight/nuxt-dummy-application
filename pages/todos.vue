<template>
  <section class="section">
    <ul>
      <li v-for="(item, index) in list" 
        :key="index">
        {{ item }}
      </li>
    </ul>

    <el-input></el-input>
    <el-button @click="submit()">Submit</el-button>
  </section>
</template>

<script>
'use strict'

import {
  Input,
  Button
} from 'element-ui'

import {
  mapState,
  mapActions
} from 'vuex'

export default {
  name: 'todos-view',

  components: {
    'el-input': Input,
    'el-button': Button
  },

  data () {
    return {
      // local state
    }
  },

  computed: {
    ...mapState('todos', {
      list: state => state.list
    })
  },

  methods: {
    ...mapActions('todos', [
      'addToDo'
    ]),

    submit () {
      this.addToDo({
        text: 'example state'
      })
    }
  }
}
</script>

<style lang="scss" scoped>
ul {
  height: auto;
}

li {
  line-height: 1;
  height: 1em;
}
</style>