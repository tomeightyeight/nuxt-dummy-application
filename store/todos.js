export const state = () => ({
  list: [
    'first example item',
    'second example item'
  ]
})

export const types = {
  ADD_TODO: 'ADD_TODO',
  REMOVE_TODO: 'REMOVE_TODO'
}

export const mutations = {
  [types.ADD_TODO] (state, { text }) {
    state.list.push(text)
  },

  [types.REMOVE_TODO] (state, { index }) {
    state.list.splice(index, 1)
  }
}

export const actions = {
  addToDo ({ commit }, text) {
    commit({
      type: types.ADD_TODO,
      payload: {
        text: text
      }
    })
  }
}
