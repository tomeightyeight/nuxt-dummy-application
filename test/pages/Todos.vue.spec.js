/* eslint-disable */
'use strict'

import { shallow, mount, createLocalVue } from 'vue-test-utils'
import { createRenderer } from 'vue-server-renderer'

import Vuex from 'vuex'
import { state } from '@/store/todos.js'

import TodosView from '@/pages/todos.vue'

const localVue = createLocalVue()
localVue.use(Vuex)

describe('todos.vue', () => {
  let actions
  let getters
  let store

  beforeEach(() => {

    // mock the store

    actions = {
      addToDo: jest.fn(),
      removeToDo: jest.fn(),
      fetchToDos: jest.fn()
    }

    getters = {
      completedToDos: jest.fn()
    }

    store = new Vuex.Store({
      modules: {
        todos: {
          namespaced: true,
          state,
          getters,
          actions
        }
      }
    })
  })

  it('snapshot: has same HTML structure', () => {
    const renderer = createRenderer()
    const wrapper = mount(TodosView, { store, localVue })
    
    renderer.renderToString(wrapper.vm, (err, str) => {
      if (err) {
        throw new Error(err)
      }

      expect(str).toMatchSnapshot()
    })   
  })

  it('renders li for each item in store', () => {
    const wrapper = shallow(TodosView, { store, localVue })
    const items = wrapper.findAll('li')

    expect(items).toHaveLength(store.state.todos.list.length)
  })

  it('dispatch store action addToDo when submit button clicked', () => {
    const wrapper = mount(TodosView, { store, localVue })
    const button = wrapper.find('.btn-submit')

    wrapper.setData({
      input: 'foo'
    })
    
    button.trigger('click')
    
    expect(actions.addToDo).toBeCalled()
  })

  it('does not dispatch store action addToDo if input empty', () => {
    const wrapper = mount(TodosView, { store, localVue })
    const input = wrapper.find('.input-new-todo')
    const button = wrapper.find('.btn-submit')

    wrapper.setData({
      input: ''
    })

    expect(actions.addToDo).not.toBeCalled()
  })

  it('dispatch store action removeToDo when remove button is clicked', () => {
    const wrapper = mount(TodosView, { store, localVue })       
    const button = wrapper.find('.btn-remove')
    
    button.trigger('click')

    expect(actions.removeToDo).toBeCalled()
  })

  it('dispatch store action fetchToDos when fetch button is clicked', () => {
    const wrapper = mount(TodosView, { store, localVue })  
    const button = wrapper.find('.btn-fetch')
    
    button.trigger('click')

    expect(actions.fetchToDos).toBeCalled()
  })
})
