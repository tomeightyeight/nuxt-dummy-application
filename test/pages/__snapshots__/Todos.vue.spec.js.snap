// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todos.vue snapshot: has same HTML structure 1`] = `
<section class="section">
    <ul>
        <li>
            delectus aut autem original
            <button type="button" class="el-button btn-remove el-button--default">
                <!---->
                <!----><span>X</span></button>
        </li>
        <li>
            quis ut nam facilis et original
            <button type="button" class="el-button btn-remove el-button--default">
                <!---->
                <!----><span>X</span></button>
        </li>
    </ul>
    <div class="input-new-todo el-input">
        <!---->
        <input autocomplete="off" type="text" rows="2" autoComplete="off" validateEvent="true" value="" class="el-input__inner">
        <!---->
        <!---->
        <!---->
    </div>
    <button type="button" class="el-button btn-submit el-button--default">
        <!---->
        <!----><span>Submit</span></button>
    <button type="button" class="el-button btn-fetch el-button--default">
        <!---->
        <!----><span>Fetch</span></button>
    <div><span class="some-test-class">Completed: </span></div>
</section>
`;
